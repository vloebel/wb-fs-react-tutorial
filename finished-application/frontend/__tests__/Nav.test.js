import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import waait from 'waait';
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CartStateProvider } from '../components/LocalState';

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: fakeUser() } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: {
          // This typename must be explicit because we use the `... on User` fragment
          __typename: 'User',
          ...fakeUser(),
          cart: [fakeCartItem()],
        },
      },
    },
  },
];

describe('<Nav/>', () => {
  it('renders a minimal nav when signed out', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders full nav when signed in', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider
          mocks={signedInMocks}
          defaultOptions={{ watchQuery: { fetchPolicy: 'no-cache' } }}
        >
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('Sign Out');
    expect(container).toHaveTextContent('My Cart');
  });

  it('renders the amount of items in the cart', async () => {
    render(
      <CartStateProvider>
        <MockedProvider
          mocks={signedInMocksWithCartItems}
          defaultOptions={{ watchQuery: { fetchPolicy: 'no-cache' } }}
        >
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    expect(screen.getByText('3')).toBeInTheDocument();
  });
});
