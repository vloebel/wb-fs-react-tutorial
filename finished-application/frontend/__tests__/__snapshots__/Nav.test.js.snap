// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 kIkwEG"
    data-testid="nav"
  >
    <a
      href="/products"
    >
      Shop
    </a>
    <a
      href="/signup"
    >
      Sign In
    </a>
  </ul>
</div>
`;

exports[`<Nav/> renders full nav when signed in 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 kIkwEG"
    data-testid="nav"
  >
    <a
      href="/products"
    >
      Shop
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/me"
    >
      Account
    </a>
    <button
      type="button"
    >
      Sign Out
    </button>
    <button
      type="button"
    >
      My Cart
      <span
        class="CartCount__AnimationStyles-xxvp4g-0 iqpOAB"
      >
        <div>
          <div
            class="CartCount__Dot-xxvp4g-1 eYNabh count"
          >
            0
          </div>
        </div>
      </span>
    </button>
  </ul>
</div>
`;
